### PR TITLE
feat: add support for DSCP and TTL / Hop Limit

### PR DESCRIPTION
**Pull request by crisidev**
## What does this PR do

This PR improves the support for setting and retrieving of IP headers values for

* Incoming packets IPv4 TTL (IP_RECVTTL)
* Incoming packets IPv6 Hop Limit (IPV6_RECVHOPLIMIT)
* Outgoing and incoming packets IPv4 DSCP (IP_TOS)
* Outgoing and incoming packets IPv6 DSCP (IPV6_RECVTCLASS)

It includes the socket options and the appropriate control messages for `sendmsg` and `recvmsg`.

**Note:** I am not 100% sure I understand correctly the `cfg` values and feature flags of `nix`, so they could be set wrongly in the PR.

## Checklist:

- [x] I have read `CONTRIBUTING.md`
- [x] I have written necessary tests and rustdoc comments
- [x] A change log has been added if this PR modifies nix's API

### CODE REVIEW

File: test/sys/test_socket.rs
Comment (crisidev) [inline test/sys/test_socket.rs:2638]: This `bind` call with `IPV6_RECVTCLASS` fails on the `Linux aarch64` target with `EADDRNOTAVAIL`, so I protected the `bind` call and run the tests only if it succeeds. 

This is not optimal and I am happy to hear suggestions for this.

File: src/sys/socket/mod.rs
Comment (SteveLauC) [inline src/sys/socket/mod.rs:1030]: Same here, FreeBSD uses `IP_RECVTOS`

File: src/sys/socket/mod.rs
Comment (SteveLauC) [inline src/sys/socket/mod.rs:1025]: BTW, do we have any documents on the size that should be used here, I didn't find much about this...

**Comment by SteveLauC** [comment]
> Note: I am not 100% sure I understand correctly the cfg values and feature flags of nix, so they could be set wrongly in the PR.

For the cfg for targets (OSes), see https://github.com/nix-rust/nix/blob/5fde28e209032e1714e726b3c38ac787315a6e52/build.rs#L20-L27

File: src/sys/socket/sockopt.rs
Comment (crisidev) [reply src/sys/socket/sockopt.rs:409]: Oh, I think I see what you mean.. This can be used to get the TOS from the socket itself right?

**Comment by crisidev** [comment]
> Hi, thanks for the PR! Sorry for the late response!

Thanks a lot for the review and don't worry about the timing :)

I'll address the comments and publish a new version!

> 
> > Note: I am not 100% sure I understand correctly the cfg values and feature flags of nix, so they could be set wrongly in the PR.
> 
> For the cfg for targets (OSes), see
> 
> https://github.com/nix-rust/nix/blob/5fde28e209032e1714e726b3c38ac787315a6e52/build.rs#L20-L27

File: src/sys/socket/sockopt.rs
Comment (SteveLauC) [reply src/sys/socket/sockopt.rs:409]: > This can be used to get the TOS from the socket itself using getsockopt right?

Yeah, that's what I mean:)

File: test/sys/test_socket.rs
Comment (crisidev) [reply test/sys/test_socket.rs:2638]: Done

File: src/sys/socket/sockopt.rs
Comment (crisidev) [reply src/sys/socket/sockopt.rs:409]: Done

File: src/sys/socket/sockopt.rs
Comment (crisidev) [reply src/sys/socket/sockopt.rs:421]: Done

File: src/sys/socket/sockopt.rs
Comment (crisidev) [reply src/sys/socket/sockopt.rs:443]: Done

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs:778]: Sounds good to me.

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs:784]: Sounds good to me.

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs:1025]: I did not find much as well, so I assumed a `u8` would be enough (same as v4) giving us a max of 256 hops.. What do you think?

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs:1030]: I am not 100% sure I follow here. Is the suggestion to have 2 different blocks, one for Linux and one for FreeBSD?

File: test/sys/test_socket.rs
Comment (crisidev) [reply test/sys/test_socket.rs:2549]: I think some of them (maybe all) will probably work.. I'll validate once I am done with addressing your comments. If they do not work, I'll try to document why.

**Comment by crisidev** [comment]
The code we have right now does not build on freebsd 14 because 

```
error[E0531]: cannot find unit struct, unit variant or constant `IP_RECVTTL` in crate `libc`
    --> src/sys/socket/mod.rs:1018:38
     |
1018 |             (libc::IPPROTO_IP, libc::IP_RECVTTL) => {
     |                                      ^^^^^^^^^^ help: a constant with a similar name exists: `IP_RECVTOS`
     |
    ::: /.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.155/src/unix/bsd/freebsdlike/freebsd/mod.rs:3685:1
     |
3685 | pub const IP_RECVTOS: ::c_int = 68;
     | ----------------------------- similarly named constant `IP_RECVTOS` defined here
```

I think we need to enable RECVTTL only on linux.

File: src/sys/socket/sockopt.rs
Comment (SteveLauC) [inline src/sys/socket/sockopt.rs:409]: Need a blank space here:

```suggestion
    /// Set or receive the Type-Of-Service (TOS) field that is
```

File: src/sys/socket/mod.rs
Comment (SteveLauC) [reply src/sys/socket/mod.rs]: Sorry for being unclear, on FreeBSD, this should check `IP_RECVTTL`, on Linux, `IP_TTL` should be used.

> Is the suggestion to have 2 different blocks, one for Linux and one for FreeBSD?

Yes

This is documented in [man 7 ip](https://man7.org/linux/man-pages/man7/ip.7.html):

> Some BSD sockets implementations also provide an IP_RECVTTL option, but an ancillary message with type IP_RECVTTL is passed with the incoming packet.  This is different from the IP_TTL option used in Linux.

File: src/sys/socket/mod.rs
Comment (SteveLauC) [reply src/sys/socket/mod.rs]: That's right, sorry for being unclear.

**Comment by SteveLauC** [comment]
> The code we have right now does not build on freebsd 14 because
> 
> ```
> error[E0531]: cannot find unit struct, unit variant or constant `IP_RECVTTL` in crate `libc`
>     --> src/sys/socket/mod.rs:1018:38
>      |
> 1018 |             (libc::IPPROTO_IP, libc::IP_RECVTTL) => {
>      |                                      ^^^^^^^^^^ help: a constant with a similar name exists: `IP_RECVTOS`
>      |
>     ::: /.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.155/src/unix/bsd/freebsdlike/freebsd/mod.rs:3685:1
>      |
> 3685 | pub const IP_RECVTOS: ::c_int = 68;
>      | ----------------------------- similarly named constant `IP_RECVTOS` defined here
> ```
> 
> I think we need to enable RECVTTL only on linux.

On FreeBSD, `IP_RECVTTL` should be used, [it is available on FreeBSD](https://github.com/freebsd/freebsd-src/blob/d26c565597b88eb5c6000dbba9e348debe7095e3/sys/netinet/in.h#L499), our CI says it does not exist because the Rust libc crate does not include it, we need to add this constant to the libc crate.

File: src/sys/socket/mod.rs
Comment (SteveLauC) [inline src/sys/socket/mod.rs:1196]: A change made by accident:)

File: src/sys/socket/mod.rs
Comment (SteveLauC) [reply src/sys/socket/mod.rs:1025]: FreeBSD manual has 3 flags documented:

| IP_RECVTTL | IP_RECVTOS | IP_V6_HOPLIMIT|IPV6_TCLASS |
|------------------|--------------------|-------------------------|-------------------|
| u_char         | u_char             | int|             undocumented  |

[RFC 3542 (IPV6)](https://datatracker.ietf.org/doc/html/rfc3542.html) says `IPV6_HOPLIMIT` and `IPV6_TCLASS` should both use `int`:

```
   opt level/    optname/          optval/
      cmsg_level    cmsg_type         cmsg_data[]
      ------------  ------------      ------------------------
      IPPROTO_IPV6  IPV6_PKTINFO      in6_pktinfo structure
      IPPROTO_IPV6  IPV6_HOPLIMIT     int
      IPPROTO_IPV6  IPV6_NEXTHOP      socket address structure
      IPPROTO_IPV6  IPV6_RTHDR        ip6_rthdr structure
      IPPROTO_IPV6  IPV6_HOPOPTS      ip6_hbh structure
      IPPROTO_IPV6  IPV6_DSTOPTS      ip6_dest structure
      IPPROTO_IPV6  IPV6_RTHDRDSTOPTS ip6_dest structure
      IPPROTO_IPV6  IPV6_TCLASS       int
```

For Linux, I am not sure the type to use, maybe we should take a look at the kernel code:<

Comment (SteveLauC) [reply src/sys/socket/mod.rs:1025]: Ok, I have checked the Linux kernel source code:

| IP_TTL | IP_TOS | IP_HOPLIMIT | IP_TCLASS |
|-----------|------------|---------------------|----------------|
| int        |u_char   int | int|int|

ref: 

v4: https://github.com/torvalds/linux/blob/44ef20baed8edcb1799bec1e7ad2debbc93eedd8/net/ipv4/ip_sockglue.c#L63-L72

v6: 

1. https://github.com/search?q=repo%3Atorvalds%2Flinux+path%3A%2F%5Enet%5C%2F%2F+IPV6_HOPLIMIT&type=code  
2. https://github.com/torvalds/linux/blob/44ef20baed8edcb1799bec1e7ad2debbc93eedd8/net/ipv6/ipv6_sockglue.c#L1174

**Comment by crisidev** [comment]
I am moving this back to draft, while I find some time to finish to implement the changes and fix all the tests.

**Comment by SteveLauC** [comment]
HI @crisidev, would you like to finish this? That type thing should be easy to deal with as we know the type used by the OS kernel:)

**Comment by crisidev** [comment]
Hey, sorry, life got in the middle.. I'd love to finish this, but I am finally on holidays for a couple of weeks. I will be able to restart working on this in the second week of August if that's ok..

**Comment by SteveLauC** [comment]
> but I am finally on holidays for a couple of weeks. I will be able to restart working on this in the second week of August if that's ok..

Yeah, that is totally fine, enjoying you holiday, we all have a life😁

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs]: Done.

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs]: Done.

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs:1025]: I think this is done now.

File: src/sys/socket/sockopt.rs
Comment (crisidev) [reply src/sys/socket/sockopt.rs:409]: Fixed.

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs:1196]: Done.

**Comment by crisidev** [comment]
> > but I am finally on holidays for a couple of weeks. I will be able to restart working on this in the second week of August if that's ok..
> 
> Yeah, that is totally fine, enjoy you holiday, we all have a life😁

I found some time during a lazy Sunday and I am trying to finish this 😄 

CI is making my life very hard right now, I'll try to fix it and move this to non draft.

**Comment by crisidev** [comment]
I think I am done fixing CI issues. I had to disable the tests for IPTOS / IPV6TCLASS on mips since they return bogus values and I don't have a good way to troubleshoot it.

**Comment by crisidev** [comment]
@SteveLauC first of all, thanks for all the patient comments, I really appreciate the help. I think this is ready for review now!

File: src/sys/socket/mod.rs
Comment (SteveLauC) [reply src/sys/socket/mod.rs:1242]: Seems like we should rename this variant to `Ipv4Tos`, to be consistent with other `Ipv4Xxx` stuff, e.g., `Ipv4Ttl`:)

-----

Well, I just checked the existing `ControlMessage/ControlMessageOwned` and socket options, looks like Nix does not have a convention for this, some variants/types contains `v4` in their name, and some don't. I think we should include `v4` in the name, as Rust std does this to its types, your thoughts on this?

File: src/sys/socket/sockopt.rs
Comment (SteveLauC) [inline src/sys/socket/sockopt.rs:439]:
```suggestion
#[cfg(any(linux_android, target_os = "freebsd"))]
```

File: src/sys/socket/sockopt.rs
Comment (SteveLauC) [inline src/sys/socket/sockopt.rs:1103]:
```suggestion
#[cfg(any(linux_android, target_os = "freebsd"))]
```

File: test/sys/test_sockopt.rs
Comment (SteveLauC) [inline test/sys/test_sockopt.rs:882]:
```suggestion
#[cfg(any(linux_android, target_os = "freebsd"))]
```

File: test/sys/test_sockopt.rs
Comment (SteveLauC) [inline test/sys/test_sockopt.rs:910]:
```suggestion
#[cfg(any(linux_android, target_os = "freebsd"))]
```

File: test/sys/test_sockopt.rs
Comment (SteveLauC) [inline test/sys/test_sockopt.rs:941]:
```suggestion
#[cfg(any(linux_android, target_os = "freebsd"))]
```

File: test/sys/test_sockopt.rs
Comment (SteveLauC) [inline test/sys/test_sockopt.rs:969]:
```suggestion
#[cfg(any(linux_android, target_os = "freebsd"))]
```

File: test/sys/test_socket.rs
Comment (SteveLauC) [inline test/sys/test_socket.rs:2580]: I have added the socket option `IP_TOS` and `IPV6_TCLASS` for Android and FreeBSD in #2464, so hopefully these 2 tests should work on them as well, so:

```suggestion
#[cfg(any(linux_android, target_os = "freebsd"))]
```

**Review by SteveLauC — COMMENTED** [review]
Well, sorry for so many comments, I hope they won't affect your holiday:)

Most of them are trivial `cfg` issues, so should be easy to fix

**Comment by crisidev** [comment]
> Well, sorry for so many comments, I hope they won't affect your holiday:)
> 
> Most of them are trivial `cfg` issues, so should be easy to fix

Please don't you worry about this. I am quite new to this crate and without your help, I would be able to make a meaningful contribution. At the end, you did most of the work, I am really glad this is moving forward thanks to your help. 

I'll try to fix everything today, I think it start to look good, apologies for the mess I did with `cfg` flags.

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs:798]: I was indeed confused by the `u_char int` in the table. Thanks for the clarification.

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs]: Sounds good.

File: src/sys/socket/mod.rs
Comment (crisidev) [reply src/sys/socket/mod.rs]: I sort of followed what I found in the existing codebase, where most of the v4 stuff is not prefixed with `v4`, while most of the v6 is prefixed with `v6`. I guess the reason is coming from the names in the kernel headers and in the `libc` crate. I like the clarity of having the prefix also for v4 TBH. I would be nice to have a consistent naming in the crate, but it would be a gigantic breaking change, and this is up to the maintainers to decide IMO :)

File: src/sys/socket/sockopt.rs
Comment (crisidev) [reply src/sys/socket/sockopt.rs:422]: I think so. Let's do this: I'll add the v4 prefix and we see how it look like in next revision? It's easy to come back with a rename!

File: src/sys/socket/mod.rs
Comment (crisidev) [inline src/sys/socket/mod.rs:1548]: I think you mentioned to return `libc::IP_TOS` on Linux and `libc::IP_RECVTOS` on freebsd, but the tests return `ENOPROTO` on Freebsd if we do this, while they work properly if we return `libc::IP_TOS` for both os flavours.

Comment (SteveLauC) [reply src/sys/socket/mod.rs:1554]: Just rechecked the FreeBSD manual, looks that the IP_RECVXXX is only used with recvmsg(). 

Yeah, FreeBSD should use the same cmsg_type when it comes to ControlMessage, let's update this for both Ipv4Ttl and Ipv4Tos.


Sorry for being misleading:<

**Comment by crisidev** [comment]
I think I could use another review now :)

File: test/sys/test_socket.rs
Comment (SteveLauC) [inline test/sys/test_socket.rs:2654]:
```suggestion
```

**Review by SteveLauC — APPROVED** [review]
Just one little thing, let me commit this change:)

**Comment by SteveLauC** [comment]
Thanks!

**Pull request merged**
